### DISCUSSION TITLE
Uploading zip to back-end in chunks

### DISCUSSION THREAD

**Discussion by alexandis** [Touffy/client-zip#81] [Q&A]
I have many files (total amount could be GBs), which I need to upload to back-end. I want to do that in - let's say - 300MB chunks (i.e. one API call == 1 chunk).
I've tried to use the following approach for reading the chunks of this specific size (`reader.read({ buffer: chunk, byteLength: 300 * 1024 * 1024, byteOffset: offset })`):

```
      const filesWithMeta: InputWithMeta[] = files.map(file => ({ input: file, name: file.name.replace(/^\/+/, '') }));
      const zipStream = makeZip(filesWithMeta);
      const reader = zipStream.getReader({ mode: 'byob' });

```
But I am getting the exception. When I use an ordinary `getReader`, it does work, but the logic is getting pretty complex and very slow (when accumulating the read bytes inside `Uint8Array` until it's getting of chunk size for upload), much slower than **JsZip** approach. Am I missing something?

**Comment by Touffy** [comment]
Your code looks right at first glance. What exception do you get ?

I actually want to switch to BOYB streams inside client-zip, but I'm waiting until Safari adds support.

(do you really need to zip the files to send them to your back-end ? and why not send the whole archive in one API call ?)

**Comment by alexandis** [comment]
The exception I'm getting is specially about 'byob'  mode unsupported (I didn't save it, since now trying different approaches).

And I cannot send a whole file, since there's limitation per request at our server. On the other hand, there could be a lot of small files, so I don't want to have to many upload requests. That's how I came up with the idea of uniform chunks.

I'm even not sure now I was moving in right direction and 'byob' would really help me. What I was expecting was reading the resulted zip file in chunks - where each chunk size would be predefined - so I would direct this chunk straight to upload.

**Reply by Touffy** [reply]
I don't know how much overhead you have for sending each request, but you should try to measure it. It may be less than you think, especially if the server supports http/2 (http/2 does not need a new TCP connection for each new request).

It's possible that you simply can't get a BYOB reader on a default stream (not created with `type: 'bytes'`), and client-zip's output is a default stream.

Anyway, you don't need to accumulate data into a 300 MB buffer to send it in roughly 300 MB chunks. Each request can still be streamed smoothly from client-zip's output into the `fetch` request body. I don't know if there's already a library implementing the functions to do that easily, though. Implementing from scratch just with vanilla ReadableStreams could be a little tedious.
